base/src/main/java/io/spine/code/proto/PackageName.java

```diff
@@ -84,4 +93,96 @@ public boolean isInnerOf(PackageName parentCandidate) {
         boolean result = value().startsWith(parentCandidate.value());
         return result;
     }
+
+    /**
+     * Appends the specified package name as if it is the child of this one.
+     *
+     * <p>For example:
+     *
+     * <pre>
+     *     {@code
+     *     PackageName enrichments = new PackageName("enrichment");
+     *
+     *     return new PackageName("spine.type").resolve(enrichments);
+     *     }
+     * </pre>
+     *
+     * returned value is a new package name, referencing a {@code "spine.type.enrichments"}
+     * Protobuf package.
+     *
+     * @return a new package name, constructed from the specified one being appended to this
+     *         package name
+     */
+    public PackageName resolve(PackageName other) {
+        return new PackageName(this.value() + DELIMITER + other.value());
+    }
+
+    /**
+     * Tests whether this package has the specified one as a child
+     */
+    public boolean contains(PackageName childCandidate) {
+        checkNotNull(childCandidate);
+        boolean result = value().contains(childCandidate.value());
+        return result;
+    }
+
+    /**
+     * Tests whether the specified package is reachable from this one.
+     *
+     * <p>Packages are reachable either when they are nested, or if their names explicitly
+     * reflect their difference.
+     *
+     * <p>For example, a package with name {@code "spine.type.user"} can be reached from the package
+     * {@code "spine"} if the specified name is {@code "type.user"}.
+     */
+    public boolean canReach(PackageName name) {
+        // Packages referenced by their FQN are always reachable.
+        if (name.isFqn()) {
+            return true;
+        }
+        return childExists(name);
+    }
+
+    private boolean childExists(PackageName child) {
+        ImmutableSet<PackageName> allPackages = KnownTypes.instance()
+                                                          .packageNames();
+        ImmutableList<PackageName> subpackages = subpackages();
+        boolean result = subpackages.stream()
+                                    .map(thisRef -> thisRef.resolve(child))
+                                    .anyMatch(allPackages::contains);
+        return result;
+    }
+
+    /**
+     * Tests whether this package name is fully qualified.
+     */
+    public boolean isFqn() {
+        boolean result = KnownTypes.instance()
+                                   .packageNames()
+                                   .contains(this);
+        return result;
+    }
+
+    /**
+     * Obtains all subpackages of this package.
+     *
+     * <p>A subpackage is a package nested in the current one.
+     *
+     * <p>A package is always its own subpackage.
+     *
+     * <p>For example, for a package {@code spine.type.user.customers} subpackages are exactly
+     * {@code [spine, spine.type, spine.type.user, spine.type.user.customers]}.
+     */
+    public ImmutableList<PackageName> subpackages() {
```

Comment:
We had an API which speaks about nesting in terms of `inner`. This is how Protobuf documentation refers to them.

Please also see my comment about having a graph.